plugins/bundle/plugin_test.go

```diff
@@ -107,6 +107,66 @@ func TestPluginStart(t *testing.T) {
 	}
 }
 
+func TestStop(t *testing.T) {
+	var longPollTimeout int64 = 3
+	tsURLBase := "/opa-test/"
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !strings.HasPrefix(r.URL.Path, tsURLBase) {
+			t.Fatalf("Invalid request URL path: %s, expected prefix %s", r.URL.Path, tsURLBase)
+		}
+
+		// simulate long operation
+		time.Sleep(time.Duration(longPollTimeout) * time.Second)
+		fmt.Fprintln(w, "") // Note: this is an invalid bundle and will fail the download
+	}))
+	defer ts.Close()
+
+	ctx := context.Background()
+	manager := getTestManager()
+
+	serviceName := "test-svc"
+	err := manager.Reconfigure(&config.Config{
+		Services: []byte(fmt.Sprintf("{\"%s\":{ \"url\": \"%s\"}}", serviceName, ts.URL+tsURLBase)),
+	})
+	if err != nil {
+		t.Fatalf("Error configuring plugin manager: %s", err)
+	}
+
+	baseConf := download.Config{Polling: download.PollingConfig{LongPollingTimeoutSeconds: &longPollTimeout}}
+
+	plugin := Plugin{
+		manager:     manager,
+		status:      map[string]*Status{},
+		etags:       map[string]string{},
+		downloaders: map[string]bundleLoader{},
+	}
+	bundleName := "test-bundle"
+	plugin.status[bundleName] = &Status{Name: bundleName}
+
+	callback := func(ctx context.Context, u download.Update) {
+		plugin.oneShot(ctx, bundleName, u)
+	}
+	plugin.downloaders[bundleName] = download.New(baseConf, plugin.manager.Client(serviceName), bundleName).WithCallback(callback)
+
+	err = plugin.Start(ctx)
+	if err != nil {
+		t.Fatal("unexpected error:", err)
+	}
+
+	// Give time for a long poll request to be initiated
+	time.Sleep(1 * time.Second)
```

Comment:
Is there any easy way to avoid this? Could we somehow check in `ts`'s handler that this has happened...?

Reply:
What if we create a channel and close it `ts`'s handler ? So we'd replace `time.Sleep(1 * time.Second)` by `<-done` which would block till the handler is invoked. This seems to work in my testing.  WDYT ?

Reply:
Sounds good!

Reply:
Done